test/integration/http_integration.cc

```diff
@@ -635,6 +635,121 @@ void HttpIntegrationTest::testRetry() {
+  EXPECT_TRUE(response->complete());
+  EXPECT_STREQ("408", response->headers().Status()->value().c_str());
+  EXPECT_EQ("stream timeout", response->body());
+}
+
+// Per-stream idle timeout after upstream headers have been sent.
+void HttpIntegrationTest::testPerStreamIdleTimeoutAfterUpstreamHeaders() {
+  auto response = setupPerStreamIdleTimeoutTest();
+
+  upstream_request_->encodeHeaders(Http::TestHeaderMapImpl{{":status", "200"}}, false);
+
+  if (downstream_protocol_ == Http::CodecClient::Type::HTTP1) {
+    codec_client_->waitForDisconnect();
+  } else {
+    response->waitForReset();
+    codec_client_->close();
+  }
+
+  EXPECT_FALSE(upstream_request_->complete());
+  EXPECT_EQ(0U, upstream_request_->bodyLength());
+
+  EXPECT_EQ(1, test_server_->counter("http.config_test.downstream_rq_idle_timeout")->value());
+  EXPECT_FALSE(response->complete());
+  EXPECT_STREQ("200", response->headers().Status()->value().c_str());
+  EXPECT_EQ("", response->body());
+}
+
+// Per-stream idle timeout after a sequence of header/data events.
+void HttpIntegrationTest::testPerStreamIdleTimeoutAfterData() {
+  auto response = setupPerStreamIdleTimeoutTest();
+
+  upstream_request_->encodeHeaders(Http::TestHeaderMapImpl{{":status", "200"}}, false);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(600));
+  upstream_request_->encodeData(1, false);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(600));
+  codec_client_->sendData(*request_encoder_, 1, false);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(600));
+  Http::TestHeaderMapImpl request_trailers{{"request1", "trailer1"}, {"request2", "trailer2"}};
+  codec_client_->sendTrailers(*request_encoder_, request_trailers);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(600));
+  upstream_request_->encodeData(1, false);
+
+  if (downstream_protocol_ == Http::CodecClient::Type::HTTP1) {
+    codec_client_->waitForDisconnect();
+  } else {
+    response->waitForReset();
+    codec_client_->close();
+  }
+
+  EXPECT_TRUE(upstream_request_->complete());
+  EXPECT_EQ(1U, upstream_request_->bodyLength());
+
+  EXPECT_EQ(1, test_server_->counter("http.config_test.downstream_rq_idle_timeout")->value());
+  EXPECT_FALSE(response->complete());
+  EXPECT_STREQ("200", response->headers().Status()->value().c_str());
+  EXPECT_EQ("aa", response->body());
+}
+
+// Successful request/response when per-stream idle timeout is configured.
+void HttpIntegrationTest::testPerStreamIdleTimeoutRequestAndResponse() {
+  config_helper_.addConfigModifier(
+      [&](envoy::config::filter::network::http_connection_manager::v2::HttpConnectionManager& hcm)
+          -> void {
+        auto* route_config = hcm.mutable_route_config();
+        auto* virtual_host = route_config->mutable_virtual_hosts(0);
+        auto* route = virtual_host->mutable_routes(0)->mutable_route();
+        route->mutable_idle_timeout()->set_seconds(1);
+      });
```

Comment:
Let's define this once and call it from both places, a la setAllowHttp10WithDefaultHost
Alternately I don't think it'd be a bad one to add into the config utils.